runners/core-java/src/main/java/org/apache/beam/runners/core/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,517 @@
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public Coder<Checkpoint<T>> getCheckpointMarkCoder() {
+      return new CheckpointCoder<>(boundedSource.getDefaultOutputCoder());
+    }
+
+    @VisibleForTesting
+    static class Checkpoint<T> implements UnboundedSource.CheckpointMark {
+      private final @Nullable List<TimestampedValue<T>> residualElements;
+      private final @Nullable BoundedSource<T> residualSource;
+
+      public Checkpoint(
+          @Nullable List<TimestampedValue<T>> residualElements,
+          @Nullable BoundedSource<T> residualSource) {
+        this.residualElements = residualElements;
+        this.residualSource = residualSource;
+      }
+
+      @Override
+      public void finalizeCheckpoint() {}
+
+      @VisibleForTesting
+      @Nullable List<TimestampedValue<T>> getResidualElements() {
+        return residualElements;
+      }
+
+      @VisibleForTesting
+      @Nullable BoundedSource<T> getResidualSource() {
+        return residualSource;
+      }
+    }
+
+    @VisibleForTesting
+    static class CheckpointCoder<T> extends StandardCoder<Checkpoint<T>> {
+
+      @JsonCreator
+      public static CheckpointCoder<?> of(
+          @JsonProperty(PropertyNames.COMPONENT_ENCODINGS)
+          List<Coder<?>> components) {
+        Preconditions.checkArgument(components.size() == 1,
+            "Expecting 1 components, got " + components.size());
+        return new CheckpointCoder<>(components.get(0));
+      }
+
+      private final Coder<List<TimestampedValue<T>>> elemsCoder;
+      private final Coder<T> elemCoder;
+      @SuppressWarnings("rawtypes")
+      private final Coder<BoundedSource> sourceCoder =
+          NullableCoder.of(SerializableCoder.of(BoundedSource.class));
+
+      CheckpointCoder(Coder<T> elemCoder) {
+        this.elemsCoder = NullableCoder.of(
+            ListCoder.of(TimestampedValue.TimestampedValueCoder.of(elemCoder)));
+        this.elemCoder = elemCoder;
+      }
+
+      @Override
+      public void encode(Checkpoint<T> value, OutputStream outStream, Context context)
+          throws CoderException, IOException {
+        Context nested = context.nested();
+        elemsCoder.encode(value.residualElements, outStream, nested);
+        sourceCoder.encode(value.residualSource, outStream, nested);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Checkpoint<T> decode(InputStream inStream, Context context)
+          throws CoderException, IOException {
+        Context nested = context.nested();
+        return new Checkpoint<>(
+            elemsCoder.decode(inStream, nested),
+            sourceCoder.decode(inStream, nested));
+      }
+
+      @Override
+      public List<Coder<?>> getCoderArguments() {
+        return Arrays.<Coder<?>>asList(elemCoder);
+      }
+
+      @Override
+      public void verifyDeterministic() throws NonDeterministicException {
+        sourceCoder.verifyDeterministic();
```

Comment:
you can just throw here -- we know that Java serialization is never deterministic. Or fine to leave it the same, because maybe we'll switch source coders later. *sigh* up to you ;)